ydb/library/yql/minikql/comp_nodes/mkql_grace_join_imp.cpp

```diff
@@ -1111,24 +1108,97 @@ bool TTable::NextJoinedData( TupleData & td1, TupleData & td2) {
 
 void TTable::Clear() {
 
-        for (ui64 bucket = 0; bucket < NumberOfBuckets; bucket++) {
-            TTableBucket & tb = TableBuckets[bucket];
-            tb.KeyIntVals.clear();
-            tb.DataIntVals.clear();
-            tb.StringsOffsets.clear();
-            tb.StringsValues.clear();
-            tb.InterfaceValues.clear();
-            tb.InterfaceOffsets.clear();
-            tb.JoinIds.clear();
-            tb.RightIds.clear();
-
-            TTableBucketStats & tbs = TableBucketsStats[bucket];
-            tbs.TuplesNum = 0;
-            tbs.KeyIntValsTotalSize = 0;
-            tbs.StringValuesTotalSize = 0;
+    for (ui64 bucket = 0; bucket < NumberOfBuckets; bucket++) {
+        TTableBucket & tb = TableBuckets[bucket];
+        tb.KeyIntVals.clear();
+        tb.DataIntVals.clear();
+        tb.StringsOffsets.clear();
+        tb.StringsValues.clear();
+        tb.InterfaceValues.clear();
+        tb.InterfaceOffsets.clear();
+        tb.JoinIds.clear();
+        tb.RightIds.clear();
+
+        TTableBucketStats & tbs = TableBucketsStats[bucket];
+        tbs.TuplesNum = 0;
+        tbs.KeyIntValsTotalSize = 0;
+        tbs.StringValuesTotalSize = 0;
+    }
+}
+
+void TTable::InitializeBucketSpillers(ISpiller::TPtr spiller) {
+    for (size_t i = 0; i < NumberOfBuckets; ++i) {
+        TableBucketsSpillers.emplace_back(spiller, 5_MB);
+    }
+}
+
+ui64 TTable::GetSizeOfBucket(ui64 bucket) const {
+    return TableBuckets[bucket].KeyIntVals.size() * sizeof(ui64)
+    + TableBuckets[bucket].DataIntVals.size() * sizeof(ui64)
+    + TableBuckets[bucket].StringsValues.size()
+    + TableBuckets[bucket].StringsOffsets.size() * sizeof(ui32)
+    + TableBuckets[bucket].InterfaceValues.size()
+    + TableBuckets[bucket].InterfaceOffsets.size() * sizeof(ui32);
+}
+
+bool TTable::TryToReduceMemory() {
```

Comment:
I would not expect to have a bool return value for function named like this. Maybe rename it to `TryToReduceMemoryAndWait` something else?